rclpy/rclpy/task.py

```diff
@@ -236,7 +237,10 @@ def __call__(self):
             if inspect.iscoroutine(self._handler):
                 # Execute a coroutine
                 try:
-                    self._handler.send(None)
+                    if self._future is None or self._future.done():
+                        if self._future and self._future.exception():
+                            raise self._future.exception()
+                        self._future = self._handler.send(None)
```

Comment:
I don't think this can assume a coroutine will return a future. It seems like the behavior is specific to the async library being used (see https://github.com/ros2/rclpy/issues/962#issuecomment-1179457414). This change is interesting though because it seems like it would both satisfy an `rclpy.Future` by calling `send(None)` until it raised `StopIteration`, and would only call `asyncio.Future.__await__` once.

It still needs a way to wake the executor when an asyncio future completes, and given that `asyncio` isn't thread safe I bet there would be problems if a multithreaded executor was used.